inc/veutil/qt/ve_qitems_mqtt.hpp

```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+#include <veutil/qt/ve_qitem.hpp>
+
+#include <QMqttClient>
+
+#include <QHostAddress>
+#include <QTimer>
+#include <QSet>
+#include <QString>
+#include <QByteArray>
+#include <QUrl>
+
+#ifdef MQTT_WEBSOCKETS_ENABLED
+#include <QWebSocket>
+#include <QWebSocketProtocol>
+#include <QIODevice>
+class WebSocketDevice;
+#endif // MQTT_WEBSOCKETS_ENABLED
+
+class VeQItemMqttProducer;
+
+// Implementation of a VeQItem over MQTT
+class VeQItemMqtt : public VeQItem
+{
+	Q_OBJECT
+
+public:
+	VeQItemMqtt(VeQItemMqttProducer *producer);
+
+	int setValue(QVariant const &value) override;
+
+protected:
+	void setParent(QObject *parent) override;
+
+private:
+	VeQItemMqttProducer *mqttProducer() const;
+};
+
+class VeQItemMqttProducer : public VeQItemProducer
+{
+	Q_OBJECT
+	Q_PROPERTY(ConnectionState connectionState READ connectionState NOTIFY connectionStateChanged)
+	Q_PROPERTY(QMqttClient::ClientError error READ error NOTIFY errorChanged)
+	Q_PROPERTY(QString portalId READ portalId WRITE setPortalId NOTIFY portalIdChanged)
+
+public:
+
+	enum ConnectionState {
+		Idle,
+		Connecting,
+		Connected,
+		Ready,
+		Disconnected,
+		Reconnecting,
+		Failed
+	};
+	Q_ENUM(ConnectionState)
+
+	VeQItemMqttProducer(VeQItem *root, const QString &id, const QString &clientIdPrefix, QObject *parent = nullptr);
+
+	VeQItem *createItem() override;
+
+	void open(const QHostAddress &host, int port);
+
+#ifdef MQTT_WEBSOCKETS_ENABLED
+	void open(
+		const QUrl &url,
+		QMqttClient::ProtocolVersion protocolVersion);
+#endif // MQTT_WEBSOCKETS_ENABLED
+
+	QMqttClient *mqttConnection() const;
+	void setCredentials(const QString &username, const QString &password);
+	bool publishValue(const QString &uid, const QVariant &value);
+	bool readValue(const QString &uid);
```

Comment:
const?

Reply:
it's not actually const, it publishes a request to read the specified value to the broker.  Maybe I should rename it to "requestValue" ?

Reply:
sure